src/ir/ir_traversing_visitor.h

```diff
@@ -24,47 +26,89 @@ namespace raksha::ir {
 // A visitor that also traverses the children of a node and allows performing
 // different actions before (PreVisit) and after (PostVisit) the children are
 // visited. Override any of the `PreVisit` and `PostVisit` methods as needed.
-template <typename Derived>
-class IRTraversingVisitor : public IRVisitor<Derived> {
+template <typename Derived, typename Result = Unit>
+class IRTraversingVisitor : public IRVisitor<Derived, Result> {
  public:
   virtual ~IRTraversingVisitor() {}
 
+  // Gives a default value for all 'PreVisit's to start with.
+  // Should be over-ridden if the Result is not default constructable.
+  virtual Result GetDefaultValue() { return Result(); }
+
   // Invoked before all the children of `module` is visited.
-  virtual void PreVisit(const Module& module) {}
+  virtual Result PreVisit(const Module& module) {
+    return std::move(GetDefaultValue());
+  }
   // Invoked after all the children of `module` is visited.
-  virtual void PostVisit(const Module& module) {}
+  virtual Result PostVisit(const Module& module, Result in_order_result) {
+    return std::move(in_order_result);
+  }
+  // Used to accumulate child results from the Module's children.
+  // Should discard or merge `child_result` into the `accumulator`.
+  virtual Result FoldResult(Result accumulator, const Module& module,
+                            Result child_result) {
+    return std::move(accumulator);
```

Comment:
It looks like we may not need `std::move` here and in almost all the other places according to https://abseil.io/tips/77